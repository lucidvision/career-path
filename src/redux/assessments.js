const SET_ASSESSMENT = 'SET_ASSESSMENT'

export function setAssessment (assessment) {
  return {
    type: SET_ASSESSMENT,
    assessment
  }
}

const initialState = {
  assessment: []
}

export default function assessments (state = initialState, action) {
  switch (action.type) {
    case SET_ASSESSMENT :
      return {
        ...state,
        assessment: action.assessment
      }
    default :
      return state
  }
}
