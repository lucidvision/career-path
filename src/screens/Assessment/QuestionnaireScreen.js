import React, { Component } from 'react'
import { Alert, AsyncStorage, StyleSheet, Text, View } from 'react-native'
import { connect } from 'react-redux'
import { AnswerButton } from 'components'
import { setAssessment } from 'redux/assessments'
import { jobs } from 'data/jobs'
import _ from 'lodash'

const questions = [
  'Do you like people?',
  'Do you like cleaning?',
  'Do you like studying?'
]

class QuestionnaireScreen extends Component {
  static navigationOptions = {
    title: 'Questionnaire'
  }
  state = {
    currentQuestion: 0,
    answeredQuestions: []
  }
  handleAnswerPressed = (answer) => {
    const currentQuestion = this.state.currentQuestion + 1
    const answeredQuestions = [...this.state.answeredQuestions, answer]
    this.setState({
      currentQuestion,
      answeredQuestions
    })
    if (currentQuestion === questions.length) {
      this.handleCalculateResult(answeredQuestions)
    }
  }
  handleCalculateResult = async (answeredQuestions) => {
    const result = _.reduce(jobs, (acc, job) => {
      if (answeredQuestions[job.score] === 1) {
        acc.push(job)
      }
      return acc
    }, [])
    const resultString = JSON.stringify(result)
    try {
      await AsyncStorage.setItem('assessmentResult', resultString)
    } catch (error) {
      Alert(error)
    }
    this.setState({currentQuestion: 0, answeredQuestions: []})
    this.props.dispatch(setAssessment(result))
    this.props.navigation.navigate('Result', {name: 'Results', jobs: result})
  }
  render () {
    return (
      <View style={styles.container}>
        <Text style={[styles.text, styles.verticleSpacing]}>{questions[this.state.currentQuestion]}</Text>
        <View>
          <AnswerButton
            style={styles.verticleSpacing}
            onPress={() => this.handleAnswerPressed(1)}
            title='Agree' />
          <AnswerButton
            style={styles.verticleSpacing}
            onPress={() => this.handleAnswerPressed(2)}
            title='Neutral' />
          <AnswerButton
            style={styles.verticleSpacing}
            onPress={() => this.handleAnswerPressed(3)}
            title='Disagree' />
        </View>
        <View style={styles.verticleSpacing}>
          <Text style={styles.text}>Questions Left: {questions.length - this.state.currentQuestion}</Text>
        </View>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    padding: 10
  },
  verticleSpacing: {
    marginBottom: 20
  },
  text: {
    textAlign: 'center'
  }
})

export default connect()(QuestionnaireScreen)
