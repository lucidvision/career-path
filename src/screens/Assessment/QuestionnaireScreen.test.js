import React from 'react'
import { shallow } from 'enzyme'
import { QuestionnaireScreen } from './QuestionnaireScreen'
import { setAssessment } from 'redux/assessments'

const result = [
  {
    educationRequired: 'Bachelors',
    name: 'Therapist',
    salary: 50000,
    score: 0
  },
  {
    educationRequired: 'High School',
    name: 'Tour Guide',
    salary: 30000,
    score: 0
  },
  {
    educationRequired: 'Bachelors',
    name: 'Agent',
    salary: 50000,
    score: 0
  }
]

const createQuestionnaireScreen = (override) => {
  const props = {...override}

  return shallow(<QuestionnaireScreen {...props} />)
}

describe('QuestionnaireScreen', () => {
  const component = createQuestionnaireScreen()

  it('renders correctly', () => {
    expect(component).toMatchSnapshot()
  })

  afterEach(() => {
    component.setState({
      currentQuestion: 0,
      answeredQuestions: []
    })
  })

  describe('handleAnswerPressed', () => {
    it('sets state with the answer', () => {
      component.find('AnswerButton').at(0).props().onPress()
      expect(component.state()).toEqual({
        currentQuestion: 1,
        answeredQuestions: [1]
      })
    })
  })

  describe('handleCalculateResult', () => {
    const mockSetItem = jest.fn(() => {
      return Promise.resolve('success')
    })
    jest.mock('AsyncStorage', () => {
      return {
        setItem: mockSetItem
      }
    })

    const dispatchMock = jest.fn()
    const navMock = jest.fn()
    const component = createQuestionnaireScreen({
      dispatch: dispatchMock,
      navigation: {navigate: navMock}
    })
    component.setState({currentQuestion: 2})
    component.find('AnswerButton').at(0).props().onPress()

    it('saves result to storage', () => {
      expect(mockSetItem).toHaveBeenCalled()
    })

    it('resets the state', done => {
      setTimeout(() => {
        expect(component.state()).toEqual({
          currentQuestion: 0,
          answeredQuestions: []
        })
        done()
      })
    })

    it('dispatches results', done => {
      setTimeout(() => {
        expect(dispatchMock).toHaveBeenCalledWith(setAssessment(result))
        done()
      })
    })

    it('navigates to results screen', done => {
      setTimeout(() => {
        expect(navMock).toHaveBeenCalledWith('Result', {name: 'Results', jobs: result})
        done()
      })
    })
  })
})
