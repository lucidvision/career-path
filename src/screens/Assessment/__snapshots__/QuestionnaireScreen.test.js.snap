// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionnaireScreen renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
      "padding": 10,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        Object {
          "marginBottom": 20,
        },
      ]
    }
  >
    Do you like people?
  </Text>
  <View>
    <AnswerButton
      onPress={[Function]}
      style={
        Object {
          "marginBottom": 20,
        }
      }
      title="Agree"
    />
    <AnswerButton
      onPress={[Function]}
      style={
        Object {
          "marginBottom": 20,
        }
      }
      title="Neutral"
    />
    <AnswerButton
      onPress={[Function]}
      style={
        Object {
          "marginBottom": 20,
        }
      }
      title="Disagree"
    />
  </View>
  <View
    style={
      Object {
        "marginBottom": 20,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      Questions Left: 
      3
    </Text>
  </View>
</View>
`;
